### PR TITLE
Add centered layout support

### PR DESCRIPTION
**Pull request by alygin**
This PR implements the Centered Layout feature (#4685):
- Added the `toggle centered layout` action.
- The centered layout mode only takes effect when there's a single central pane.
- The state of the centered layout toggle is saved / restored between Zed restarts.
- The paddings are controlled by the `centered_layout` setting:

```json
"centered_layout": {
  "left_padding": 0.2,
  "right_padding": 0.2
}
```

This allows us to support both the VSCode-style (equal paddings) and IntelliJ-style (only left padding in Zen mode).

Release Notes:

- Added support for Centered Layout ([#4685](https://github.com/zed-industries/zed/pull/9754)).

https://github.com/zed-industries/zed/assets/2101250/2d5b2a16-c248-48b5-9e8c-6f1219619398

Related Issues:

- Part of #4382

### CODE REVIEW

**Comment by mrnugget** [comment]
Hey! Thanks for the PR. I do think we need some design input here. For example: I think we need borders around the centered layout. I also am not sure about the tab bar being this narrow.

@iamnbutler is back in a few weeks as our main designer, so I'm wondering whether we shouldn't wait until he has taken a look here. What do you think?

**Comment by alygin** [comment]
Hi Thorsten!

> @iamnbutler is back in a few weeks as our main designer, so I'm wondering whether we shouldn't wait until he has taken a look here. What do you think?

Yes, sure, there's no rush here. Let's wait for Nate.

> I think we need borders around the centered layout. I also am not sure about the tab bar being this narrow.

I checked two other implementations when I was working on this feature &mdash; VSCode and IDEA.

VSCode uses borders and shrinks tab bar to the size of the editor area in its Centered Layout and Zen modes:

vscode-centered-view

IDEA does the opposite:

idea-dstr-free

---

My personal preferences are to: 
- Not have borders because they add visual noise and don't help me much.
- Make tab bar narrow because I don't want to constantly move my eyes here and there. This is why I switch to the Centered Layout in the first place &mdash; to keep my focus of vision narrow, expecially on wide screens.

There's another effect &mdash; the buffer and project search/replace panels are also shrinked. Though it all works the same way as when the user splits panes vertically. Both the tab bar and search/replace panels shrink accordingly.

VSCode and IDEA look and behave differently here, I don't know what most users prefer, so I'm ready to make it the way you think is more in line with the Zed approach.

**Comment by mikayla-maki** [comment]
Personally, I think the VSCode small border looks a little neater. Like things are in their place. We could make it themable though, so you can override it to make it transparent.

**Comment by alygin** [comment]
> We could make it themable though, so you can override it to make it transparent.

In VSCode it's `editorGroup.border`. I didn't find its analogue in Zed.

**Comment by mikayla-maki** [comment]
You can make a new one!

**Comment by alygin** [comment]
> We could make it themable though, so you can override it to make it transparent.

In VSCode it's `editorGroup.border`. I didn't find its analogue in Zed.

**Comment by alygin** [comment]
> You can make a new one!

Done. Moved it to a separate PR to keep things focused: #9986

**Comment by alygin** [comment]
I've added borders. The color is `pane_group.border`, which defaults to `border` if not specified in the theme.

zed-centered-layout-borders

**Comment by as-cii** [comment]
Hey @alygin, nice work as usual! I tested this out manually and it seems to be working great! Would you mind fixing merge conflicts so that we can merge this?

**Comment by alygin** [comment]
Hi @as-cii! Sure, fixed.

**Comment by mikayla-maki** [comment]
Thank you!

**Comment by SomeoneToIgnore** [comment]
Hmm, looks like the terminal scrolling had been touched by this PR.
When I position the terminal on the bottom and run the command the scrolls it, it breaks.

Starting point: 
Screenshot 2024-04-17 at 10 08 12

Current `main`:
Screenshot 2024-04-17 at 10 08 15

With the PR reverted:
Screenshot 2024-04-17 at 10 08 15

I clearly see that the terminal pane is pushed down a bit, that that bit apparently is not handled well on the terminal side (??), not allowing it to be scrolled up.

**Comment by as-cii** [comment]
I think we should revert this for the time being then to prevent it from going to Preview later today. @alygin: would you mind opening a new PR that fixes these issues?

File: docs/src/configuring_zed.md [hunk 1]
```diff
@@ -142,6 +142,24 @@ For example, to disable ligatures for a given font you can add the following to
...
+"centered_layout": {
+  "left_padding": 0.2,
+  "right_padding": 0.2,
+}
+```
+
+**Options**
+
+The `left_padding` and `right_padding` options define the relative width of the
+left and right padding of the central pane from the workspace when the centered layout mode is activated. Valid values range is from `0` to `0.45`.
```
Comment (mmtftr) [inline docs/src/configuring_zed.md:161]: Doc says max padding is 0.45 but Workspace::MAX_PADDING is defined as 0.4 above?

**Comment by alygin** [comment]
> @alygin: would you mind opening a new PR that fixes these issues?

Yes, sure. I’ll come back later today.

**Comment by SomeoneToIgnore** [comment]
@alygin sorry for the extended messaging, but it's a release day.
I've got a small fix, creating a PR description right now: 
```
diff --git a/crates/workspace/src/workspace.rs b/crates/workspace/src/workspace.rs
index 7053ba138..1a0e9cc74 100644
--- a/crates/workspace/src/workspace.rs
+++ b/crates/workspace/src/workspace.rs
@@ -4061,7 +4061,7 @@ impl Render for Workspace {
                                     .overflow_hidden()
                                     .child(
                                         h_flex()
-                                            .h_full()
+                                            .flex_1()
                                             .when_some(paddings.0, |this, p| {
                                                 this.child(p.border_r_1())
                                             })
```

that is not a true fix, though, as I will try to explain in that description, but at least now there's no urgency to fix anything.

Welcome to investigate it deeper, if you want to, later, and thanks!

File: docs/src/configuring_zed.md (hunk 1 above)
Comment (SomeoneToIgnore) [reply docs/src/configuring_zed.md:161]: Thank you for spotting, will update the docs according to the code, but if that was the wrong move and 0.45 should be used instead, future PRs can fix that.

**Comment by domi413** [comment]
Could we add a setting to disable the `pane_group.border` for those who prefer a cleaner look?

**Comment by alygin** [comment]
> I wanted to ask: is there a possibility of configuring the project_panel, so it doesn't shift the centered layout screen when opened? In a way, that the project would rather float over the centered editor, rather than pushing it to the right

Hi @GregoryBai!
No, unfortunately, there's no such layout mode at the moment.

**Pull request merged**